tests/Basics/SupervisedBufferTest.cpp

```diff
@@ -0,0 +1,479 @@
+  }
+}
+
+TEST(SupervisedbufferTest, ReuseSupervisedBufferAccountsMemory) {
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  AqlValue firstValue;
+  AqlValue secondValue;
+
+  {
+    SupervisedBuffer supervisedBuffer(monitor);
+    Builder builder(supervisedBuffer);
+    builder.openArray();
+    builder.add(Value(std::string(1024, 'a')));
+    builder.close();
+    firstValue = AqlValue{builder.slice(), builder.size()};
+
+    monitor.increaseMemoryUsage(firstValue.memoryUsage());
+    ASSERT_GE(monitor.current(), firstValue.memoryUsage());
+  }
+  ASSERT_EQ(monitor.current(), firstValue.memoryUsage());
+
+  {
+    SupervisedBuffer supervisedBuffer(monitor);
+    Builder builder(supervisedBuffer);
+    builder.openArray();
+    builder.add(Value(std::string(2048, 'b')));
+    builder.close();
+    secondValue = AqlValue{builder.slice(), builder.size()};
+
+    monitor.increaseMemoryUsage(secondValue.memoryUsage());
+    ASSERT_GE(monitor.current(),
+              firstValue.memoryUsage() + secondValue.memoryUsage());
+  }
+  ASSERT_EQ(monitor.current(),
+            firstValue.memoryUsage() + secondValue.memoryUsage());
+
+  // drop our manual accounting for both values before destroying them
+  monitor.decreaseMemoryUsage(firstValue.memoryUsage());
+  monitor.decreaseMemoryUsage(secondValue.memoryUsage());
+  firstValue.destroy();
+  secondValue.destroy();
+  ASSERT_EQ(monitor.current(), 0);
+}
+
+TEST(SupervisedbufferTest, SupervisedBuilderGrowthAndRecycle) {
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    SupervisedBuffer supervisedBuffer(monitor);
+    Builder builder(supervisedBuffer);
+
+    // memory should be low >= size here
+    builder.openArray();
+    builder.add(Value(1));
+    builder.add(Value(2));
+    builder.add(Value(3));
+    builder.close();
+    AqlValue smallValue = AqlValue{builder.slice(), builder.size()};
+    // account the small value on the monitor. smallValue.memoryUsage()
+    // may be zero, in which case this has no effect.
+    monitor.increaseMemoryUsage(smallValue.memoryUsage());
+    std::size_t memory1 = monitor.current();
+    ASSERT_GE(memory1, builder.size());
+
+    builder.clear();
+    builder.openArray();
+    for (int i = 0; i < 200; ++i) {
+      builder.add(Value(std::string(1024, 'a')));
+    }
+    builder.close();
+    AqlValue largeValue = AqlValue{builder.slice(), builder.size()};
+    monitor.increaseMemoryUsage(largeValue.memoryUsage());
+    std::size_t memory2 = monitor.current();
+    ASSERT_GT(memory2, memory1);
+    ASSERT_GE(memory2, builder.size());
+
+    builder.clear();
+    std::size_t memory3 = monitor.current();
+    ASSERT_EQ(memory3, memory2);
+    monitor.decreaseMemoryUsage(largeValue.memoryUsage());
+    largeValue.destroy();
+    monitor.decreaseMemoryUsage(smallValue.memoryUsage());
+    smallValue.destroy();
```

Comment:
What I would ask for in this "clear" test is the following:

Create a Supervised Builder.
Add some large data in (more than 192 bytes)
Take current usage.
Then Clear()
Assert that current usage is still equal to above snapshot.
Then add half of the above data in the builder.
Assert that current usages is still equal to above snapshot.

(Idea is we recycle the buffer and not de_allocate)